File: app/src/main/java/com/dragonpav/dengine/RenderUtils.java
package com.dragonpav.dengine;

import android.opengl.GLES20;
import android.opengl.GLES30;
import android.opengl.GLSurfaceView;
import android.util.Log;

import androidx.annotation.Nullable;

import java.nio.FloatBuffer;
import java.nio.IntBuffer;

public class RenderUtils {
	public Program p;
	public Lighting lighting;
	public GLSurfaceView gsv;
	public ObjectBuffer createStaticVAO(FloatBuffer vertices, IntBuffer indices, FloatBuffer tex, FloatBuffer normalBuffer, int vertBufSize, int indexBufSize, int texBufSize, int normalBufferSize) {
		ObjectBuffer b = new ObjectBuffer();
		int[] tmpBuf = new int[1];
		GLES30.glGenVertexArrays(1, tmpBuf, 0);
		b.VAO = tmpBuf[0];
		GLES30.glBindVertexArray(b.VAO);
		GLES30.glGenBuffers(1, tmpBuf, 0);
		b.VBO = tmpBuf[0];
		GLES30.glBindBuffer(GLES30.GL_ARRAY_BUFFER, b.VBO);
		GLES30.glBufferData(GLES30.GL_ARRAY_BUFFER, vertBufSize, vertices, GLES30.GL_STATIC_DRAW);
		GLES30.glVertexAttribPointer(0, 3, GLES30.GL_FLOAT, false, 3 * 4, 0);
		GLES30.glEnableVertexAttribArray(0);
		GLES30.glGenBuffers(1, tmpBuf, 0);
		b.VBOtexture = tmpBuf[0];
		GLES30.glBindBuffer(GLES30.GL_ARRAY_BUFFER, b.VBOtexture);
		GLES30.glBufferData(GLES30.GL_ARRAY_BUFFER, texBufSize, tex, GLES30.GL_STATIC_DRAW);
		GLES30.glVertexAttribPointer(1, 2, GLES30.GL_FLOAT, false, 2 * 4, 0);
		GLES30.glEnableVertexAttribArray(1);
		if (normalBuffer != null && normalBufferSize > 0) {
			GLES30.glGenBuffers(1, tmpBuf, 0);
			b.VBOnormal = tmpBuf[0];
			GLES30.glBindBuffer(GLES30.GL_ARRAY_BUFFER, b.VBOnormal);
			GLES30.glBufferData(GLES30.GL_ARRAY_BUFFER, normalBufferSize, normalBuffer, GLES30.GL_STATIC_DRAW);
			GLES30.glVertexAttribPointer(2, 3, GLES30.GL_FLOAT, false, 3 * 4, 0);
			GLES30.glEnableVertexAttribArray(2);
		} else {
			b.VBOnormal = 0;
		}
		GLES30.glGenBuffers(1, tmpBuf, 0);
		b.EBO = tmpBuf[0];
		GLES30.glBindBuffer(GLES30.GL_ELEMENT_ARRAY_BUFFER, b.EBO);
		GLES30.glBufferData(GLES30.GL_ELEMENT_ARRAY_BUFFER, indexBufSize, indices, GLES30.GL_STATIC_DRAW);
		GLES30.glBindBuffer(GLES30.GL_ARRAY_BUFFER, 0);
		GLES30.glBindVertexArray(0);
		return b;
	}
	public RenderUtils(Program program, @Nullable Lighting li, GLSurfaceView gsv) {
		p = program;
		lighting = li;
		this.gsv = gsv;
		Log.i("DEngine", "Initializing RenderUtils...");
		Log.i("DEngine", "---------------------------");
		Log.i("DEngine", "GPU: " + GLES30.glGetString(GLES30.GL_RENDERER));
		Log.i("DEngine", "Width: " + gsv.getWidth());
		Log.i("DEngine", "Height: " + gsv.getHeight());
	}
}

File: src/RenderUtils.cpp
#include "RenderUtils.h"
#ifdef _WIN32
	RenderUtils::RenderUtils(Program* program, Lighting* li, GLFWwindow* window) {
		p = program;
		lighting = li;
		this->window = window;
		//Some logging
		std::cout << "Initializing RenderUtils..." << std::endl;
		std::cout << "------------------------" << std::endl;
		std::cout << "GPU: " << glGetString(GL_RENDERER) << std::endl;
		int width, height;
		glfwGetFramebufferSize(window, &width, &height);
		std::cout << "Width: " << width << std::endl;
		std::cout << "Height: " << height << std::endl;
	}
#else
	RenderUtils::RenderUtils(Program* program, Lighting* li, void* window) {
		p = program;
		lighting = li;
		this->window = window;
		//Some logging
		std::cout << "Initializing RenderUtils..." << std::endl;
		std::cout << "------------------------" << std::endl;
		std::cout << "GPU: " << glGetString(GL_RENDERER) << std::endl;
	}
#endif
ObjectBuffer RenderUtils::createStaticVAO(GLfloat* vertices, GLuint* indices, GLfloat* tex, GLfloat* normal, GLuint vertexBufSize, GLuint indexBufSize, GLuint texBufSize, GLuint normalBufSize) {
	ObjectBuffer b;
	//VAO
	glGenVertexArrays(1, &b.VAO);
	glBindVertexArray(b.VAO);
	//VBO
	glGenBuffers(1, &b.VBO);
	glBindBuffer(GL_ARRAY_BUFFER, b.VBO);
	glBufferData(GL_ARRAY_BUFFER, vertexBufSize, vertices, GL_STATIC_DRAW);
	glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(GLfloat), (GLvoid*)0);
	glEnableVertexAttribArray(0);
	//Texture VBO
	glGenBuffers(1, &b.VBOtexture);
	glBindBuffer(GL_ARRAY_BUFFER, b.VBOtexture);
	glBufferData(GL_ARRAY_BUFFER, texBufSize, tex, GL_STATIC_DRAW);
	glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 2 * sizeof(GLfloat), (GLvoid*)0);
	glEnableVertexAttribArray(1);
	//Normals, if lighting isn't using skipping
	if (lighting != nullptr) {
		glGenBuffers(1, &b.VBOnormal);
		glBindBuffer(GL_ARRAY_BUFFER, b.VBOnormal);
		glBufferData(GL_ARRAY_BUFFER, normalBufSize, normal, GL_STATIC_DRAW);
		glVertexAttribPointer(2, 3, GL_FLOAT, GL_FALSE, 3 * sizeof(GLfloat), (GLvoid*)0);
		glEnableVertexAttribArray(2);
	}
	//EBO
	glGenBuffers(1, &b.EBO);
	glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, b.EBO);
	glBufferData(GL_ELEMENT_ARRAY_BUFFER, indexBufSize, indices, GL_STATIC_DRAW);
	glBindBuffer(GL_ARRAY_BUFFER, 0);
	glBindVertexArray(0);
	return b;
}
#ifdef _WIN32
	void RenderUtils::setWindow(GLFWwindow* win) {
		if (window == nullptr) {
			throw std::runtime_error("RenderUtils::setWindow win is null");
			return;
		}
		window = win;
	}
#endif
void RenderUtils::setLighting(Lighting* li) {
	lighting = li;
}
void RenderUtils::setProgram(Program* prog) {
	if (p == nullptr) {
		throw std::runtime_error("RenderUtils::setProgram p is null");
		return;
	}
	p = prog;
}
#ifdef _WIN32
	GLFWwindow* RenderUtils::getWindow() {
		return window;
	}
#endif
Lighting* RenderUtils::getLighting() {
	return lighting;
}
Program* RenderUtils::getProgram() {
	return p;
}